src/lib.rs

```diff
@@ -836,50 +842,93 @@ impl<T> core::future::Future for Receiver<T> {
+            EMPTY => {
+                // SAFETY: We can't be in the forbidden states, and no waker in the channel.
+                unsafe { channel.write_async_waker(cx) }
+            }
             // We were polled again while waiting for the sender. Replace the waker with the new one.
             RECEIVING => {
+                // ORDERING: We use relaxed ordering on both success and failure since we have not
+                // written anything above that must be released, and the individual match arms
+                // handle any additional synchronization.
                 match channel
                     .state
-                    .compare_exchange(RECEIVING, EMPTY, SeqCst, SeqCst)
+                    .compare_exchange(RECEIVING, EMPTY, Relaxed, Relaxed)
                 {
                     // We successfully changed the state back to EMPTY. Replace the waker.
+                    // This is the most likely branch to be taken, which is why we don't use any
+                    // memory barriers in the compare_exchange above.
                     Ok(_) => {
+                        // SAFETY: We wrote the waker in a previous call to poll. We do not need
+                        // a memory barrier since the previous write here was by ourselves.
                         unsafe { channel.drop_waker() };
+                        // SAFETY: We can't be in the forbidden states, and no waker in the channel.
                         unsafe { channel.write_async_waker(cx) }
                     }
                     // The sender sent the message while we prepared to replace the waker.
                     // We take the message and mark the channel disconnected.
                     // The sender has already taken the waker.
                     Err(MESSAGE) => {
-                        channel.state.store(DISCONNECTED, SeqCst);
+                        // ORDERING: Synchronize with the write of the message. This branch is
+                        // unlikely to be taken.
+                        channel.state.swap(DISCONNECTED, Acquire);
+                        // SAFETY: The state tells us the sender has initialized the message.
                         Poll::Ready(Ok(unsafe { channel.take_message() }))
                     }
                     // The sender was dropped before sending anything while we prepared to park.
                     // The sender has taken the waker already.
                     Err(DISCONNECTED) => Poll::Ready(Err(RecvError)),
+                    // The sender is currently waking us up.
                     Err(UNPARKING) => {
+                        // We can't trust that the old waker that the sender has access to
+                        // is honored by the async runtime at this point. So we wake ourselves
+                        // up to get polled instantly again.
                         cx.waker().wake_by_ref();
-                        hint::spin_loop();
```

Comment:
I removed the `spin_loop`, since this will now instantly get polled again and then likely end up in the `UNPARKING` branch below, which will issue the `spin_loop` instead.

Reply:
That makes sense, looks good to me.